File: KimHyunBin/백준/0119_코테1회차_3번.java
import java.util.*;
import java.io.*;

public class Main {// Boj_22869_징검다리 건너기 (small)

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		StringBuilder sb = new StringBuilder();

		int N = Integer.parseInt(st.nextToken());
		int K = Integer.parseInt(st.nextToken());
		String answer = "";

		int[] arr = new int[N];
		boolean[] dp = new boolean[N];
		st = new StringTokenizer(br.readLine());
		for (int i = 0; i < N; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
		}
		// (j-i) * (1+Math.abs(arr[i]-arr[j])

		dp[0] = true;
		for (int i = 1; i < N; i++) {
			for (int j = 0; j <= i; j++) {
				if (dp[j]) {
					int power = (i-j) * ((1 + Math.abs(arr[i] - arr[j])));
					if (K >= power) {
						dp[i] = true;
					}
				}
			}
		}
		answer = dp[N-1] ? "YES" : "NO";
		System.out.println(answer);
	}
}

File: LeeSuJi/백준/0119_코테1회차_3번.c
// 22869 징검다리 건너기 (small)

#include <stdio.h>
#include <math.h>

int main() {
	int N, K = 0;
	scanf_s("%d %d", &N, &K);

	int A[5000] = { 0 };
	for (int i = 0; i < N; i++) {
		scanf_s("%d", &A[i]);
	}

	int result = 0; //0:실패 1:성공
	int i = 0;
	int j = i + 1;

	while (i < N - 1) {
		j = i + 1;
		while (j < N) {
			int strength = (j - i)*(1 + abs(A[i] - A[j]));
			if (strength <= K) {
				i = j;
				break;
			}
			j++;
		}
		if (i == N - 1) {
			result = 1;
			break;
		}
		else if (j == N && i != N - 1) {
			result = 0;
			break;
		}
		
	}

	char result1[4] = "YES";
	char result0[3] = "NO";

	if (result == 0)
		printf("%s", result0);
	else
		printf("%s", result1);
	return 0;
}